xarray/core/variable.py

```diff
@@ -712,6 +720,8 @@ def _getitem_with_mask(self, key, fill_value=dtypes.NA):
             data = as_indexable(self._data)[actual_indexer]
             chunks_hint = getattr(data, "chunks", None)
             mask = indexing.create_mask(indexer, self.shape, chunks_hint)
+            if isinstance(self.data, sparse.COO):
+                mask = sparse.COO.from_numpy(mask)
             data = duck_array_ops.where(mask, fill_value, data)
```

Comment:
The underlying issue is that `sparse.where` expects mask to be `SparseArray` so `create_mask` needs to know to return that. @shoyer What would be the best way to do this properly?

Reply:
Another fix is that `fill_value` should be using `self.data.fill_value` by default for a sparse array.

Reply:
It seems like we should probably pipe some information about the indexed data into `indexing.create_mask`. Right  now we do that with `chunks_hint` specifically for dask array, but that could be made more generic by passing in  `data` directly. In particular we should coerce ndarrays in the indexing key to sparse arrays inside `indexing._masked_result_drop_slice`.

Reply:
OK done. Let me know if you want more changes.